### ISSUE TITLE
Deprecate function formal parameters and function type aliases.

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/language#2901]
There are syntactical constructs that are discouraged from being used by the following lints: [use_function_type_syntax_for_parameters](https://dart-lang.github.io/linter/lints/use_function_type_syntax_for_parameters.html) as well as [prefer_generic_function_type_aliases](https://dart-lang.github.io/linter/lints/prefer_generic_function_type_aliases.html)

Because they can be replaced with other syntax e.g. with function types, I'd like to propose for function formal parameters (and related syntax) to be deprecated and eventually removed.

```
functionFormalParameter
    :    COVARIANT? type? identifier formalParameterPart '?'?
    ;

fieldFormalParameter
    :    finalConstVarOrType? THIS '.' identifier (formalParameterPart '?'?)?
    ;

superFormalParameter
    :    type? SUPER '.' identifier (formalParameterPart '?'?)?
    ;

typeAlias
    :    TYPEDEF typeIdentifier typeParameters? '=' type ';'
    |    TYPEDEF functionTypeAlias
    ;

functionTypeAlias
    :    functionPrefix formalParameterPart ';'
    ;
```

would be adjusted to:

```
fieldFormalParameter
    :    finalConstVarOrType? THIS '.' identifier
    ;

superFormalParameter
    :    type? SUPER '.' identifier
    ;

typeAlias
    :    TYPEDEF typeIdentifier typeParameters? '=' type ';'
    ;
```

i.e. 

 * `functionFormalParameter` would be removed 
 * `functionTypeAlias` would be removed
 * the optional trailing `(formalParameterPart '?'?)?` would be removed from `fieldFormalParameter` and `superFormalParameter`
 * there would only be one way to construct a type alias because there would be one type of type alias left.
 
This would:

 * simplify the grammar and therefore all related infrastructure (parsing, linting, analysis, ...).
 * make Dart simpler to learn by explicitly answering the question `which syntax should I use?`
 
Fixes are already available for both lints.

**Comment by munificent** [comment]
In a perfect world, yes, we'd remove these from Dart. However:

> * simplify the grammar and therefore all related infrastructure (parsing, linting, analysis, ...).

We've already spent the time to implement the old function type syntax and the ongoing maintenance cost is relatively low, so this doesn't really buy us much.

> * make Dart simpler to learn by explicitly answering the question which syntax should I use?

We can accomplish most of this by just telling people to use the new one and not teaching them about the old syntax.

I definitely *like* removing old deadweight features from the language. But doing so does generally break code, and that has a high cost. Unless we get significant value in return (like being able to reuse the syntax for something more useful), it's often the best choice for users to just leave it a lone.

**Comment by modulovalue** [comment]
> But doing so does generally break code, and that has a high cost.

Yes, in general, I agree. That's why I have mentioned that these features are being actively discouraged by widely used lints. This should imply that it can be expected that most users of Dart don't use those features.

I'm not necessarily asking for those constructs to be removed immediately, or anytime soon. I'm thinking more about a slow soft-deprecation period, and complete removal with e.g. Dart 4. Something along the lines of [deprecated_subtype_of_function](https://dart.dev/tools/diagnostic-messages#deprecated_subtype_of_function). This should give people enough time to migrate.

**Comment by lrhn** [comment]
You also run into people like me who *prefer*
```dart
T reduce<T>(T initialValue, T combine(T v1, T v2)) => ...
```
over
```dart
T reduce<T>(T initialValue, T Function(T, T) combine) => ...
```
I find the former vastly *easier to read*. 

Maybe it's "just" because `Function` is just too long an noisy, but that's precisely the kind of thing which affects readability.

(If we could just fix it so that a single identifier in positional parameter position would be a type, not a name, it would be perfect. But I'd rather fix than remove. The _type alias declaration_ is crap, though. Doesn't work with generic functions at all, but looks like it does. Now, if we had types on the right ... :wink: )

Edit: I've come around and gotten used to the `Function` types, and now think consistency is more valuable than the old syntax. Also because we haven't fixed it's shortcomings yet.

**Comment by modulovalue** [comment]
> I find the former vastly easier to read.

I agree that it helps to increase readability **locally** in a small example, but in my opinion, the readability argument does not apply to a constantly evolving codebase because the old syntax doesn't scale. If we want to return a function, we have to use the new syntax. I think the existence of two ways to do the same thing makes readability worse because the reader has to be ready to context-switch between alternative semantically (almost) identical syntaxes.

It's really not about syntax for me. I've gotten used to the Function syntax and have convinced myself that it is superior, but I understand that an opposing preference can be superior in other ways.

There seems to be commitment (and acceptance) from the ecosystem towards the new syntax (i.e. see the lints). I think that Dart should commit further towards the new syntax and schedule it to be removed eventually (in a year _or two or three_).

**Comment by FMorschel** [comment]
As I suggested at:

- https://github.com/dart-lang/language/issues/4730

We could consider this for dart 4.0 at least.
